torch/testing/_asserts.py

```diff
@@ -331,29 +368,230 @@ def assert_tensors_equal(
+    *,
+    check_device: bool = True,
+    check_dtype: bool = True,
+    check_stride: bool = True,
+) -> None:
+    """Asserts that the values of tensors are bitwise equal.
+
+    Optionally, checks that some attributes of tensors are equal.
+
+    Also supports :class:`~collections.abc.Sequence`'s and :class:`~collections.abc.Mapping`'s of :class:`Tensor`'s.
+
+    Args:
+        actual (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): Actual input.
+        expected (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): Expected input.
+        check_device (bool): If ``True`` (default), asserts that tensors live in the same :attr:`~torch.Tensor.device`
+            memory. If this check is disabled **and** they do not live in the same memory :attr:`~torch.Tensor.device`,
+            they are moved CPU memory before their values are compared.
+        check_dtype (bool): If ``True`` (default), asserts that tensors have the same :attr:`~torch.Tensor.dtype`. If
+            this check is disabled they do not have the same :attr:`~torch.Tensor.dtype`, they are copied to the
+            :class:`~torch.dtype` returned by :func:`torch.promote_types` before their values are compared.
+        check_stride (bool): If ``True`` (default), asserts that the tensors have the same stride.
+
+    Raises:
+        UsageError: If the input pair has an unsupported type.
+        UsageError: If any tensor is complex, quantized, or sparse. This is a temporary restriction and
+            will be relaxed in the future.
+        AssertionError: If any corresponding tensors do not have the same :attr:`~torch.Tensor.shape`.
+        AssertionError: If :attr:`check_device`, but any corresponding tensors do not live in the same
+            :attr:`~torch.Tensor.device` memory.
+        AssertionError: If :attr:`check_dtype`, but any corresponding tensors do not have the same
+            :attr:`~torch.Tensor.dtype`.
+        AssertionError: If :attr:`check_stride`, but any corresponding tensors do not have the same stride.
+        AssertionError: If the values of any corresponding tensors are not bitwise equal.
+        AssertionError: If the inputs are :class:`~collections.abc.Sequence`'s, but their length does not match.
+        AssertionError: If the inputs are :class:`~collections.abc.Mapping`'s, but their set of keys mismatch.
+
+    .. seealso::
+
+        To assert that the values in tensors are close but are not required to be bitwise equal, use
+        :func:`assert_close` instead.
+    """
+    check_tensors = functools.partial(
+        _check_tensors_equal,
+        check_device=check_device,
+        check_dtype=check_dtype,
+        check_stride=check_stride,
+    )
+    exc = _check_by_type(actual, expected, check_tensors)
     if exc:
         raise exc
 
 
-def assert_tensors_close(
-    actual: torch.Tensor,
-    expected: torch.Tensor,
+def assert_close(
+    actual: Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]],
```

Comment:
So we won't support lists of dicts whose values are lists containing tensors?

Reply:
Correct. Only top level containers are supported. Anything else will be a lot more involved to handle and I honestly can't imagine a wide-spread scenario that would warrant this.